File: rigidbodytransform/quatmath_test.py
import time
import numpy as np
from quatmath import \
    quatconj, quatinv, quatnormalize, \
    quatmultiply, quatrotate


if __name__ == '__main__':

    q = np.random.random((1,4))
    q1 = np.random.random((1,4))
    q2 = np.random.random((1,4))
    p = np.random.random((3,1))
    print('q', q)
    print('q1', q1)
    print('q2', q2)
    print('p', p)

    qconj = quatconj(q)
    print('qconj', qconj)

    qinv = quatinv(q)
    print('qinv', qinv)
    print(quatmultiply(qinv,q))

    qmultiply = quatmultiply(q1, q2)
    print('qmultiply', qmultiply)

    p1 = quatrotate(q, p)
    print('p1', p1)


    time.sleep(1)

    time_begin = time.time()
    for i in range(500000):
        q_inv = quatconj(q)
    time_end = time.time()
    print('average:', (time_end-time_begin)*1000/500000, 'ms')

    time_begin = time.time()
    for i in range(500000):
        q_inv = quatinv(q)
    time_end = time.time()
    print('average:', (time_end-time_begin)*1000/500000, 'ms')

    time_begin = time.time()
    for i in range(500000):
        q_normalized = quatnormalize(q)
    time_end = time.time()
    print('average:', (time_end-time_begin)*1000/500000, 'ms')

    time_begin = time.time()
    for i in range(500000):
        q_prod = quatmultiply(q1, q2)
    time_end = time.time()
    print('average:', (time_end-time_begin)*1000/500000, 'ms')

    time_begin = time.time()
    for i in range(500000):
        p_new = quatrotate(q, p)
    time_end = time.time()
    print('average:', (time_end-time_begin)*1000/500000, 'ms')

File: rigidbodytransform/quatmath_test.m
%%
clc;
clear;

addpath('fastquatmath');


%%
q = rand(1,4)
q1 = rand(1,4)
q2 = rand(1,4)
p = rand(3,1)

qconj = fquatconj(q)
qconj_ = quatconj(q)

qinv = fquatinv(q)
qinv_ = quatinv(q)
qequ = fquatmultiply(qinv, q)
qequ_ = quatmultiply(qinv_, q)

qmultiply = fquatmultiply(q1, q2)
qmultiply_ = quatmultiply(q1, q2)

p1 = fquatrotate(q, p)
p1_ = quatrotate(quatinv(q), p')


%%
times = 50000;


time_begin = clock;
for i=1:times
    q_inv = quatconj(q);
end
time_end = clock;
quatconj_average = etime(time_end,time_begin)*1000/times

time_begin = clock;
for i=1:times
    q_inv = fquatconj(q);
end
time_end = clock;
quatconj_average = etime(time_end,time_begin)*1000/times


time_begin = clock;
for i=1:times
    q_inv = quatinv(q);
end
time_end = clock;
quatinv_average = etime(time_end,time_begin)*1000/times

time_begin = clock;
for i=1:times
    q_inv = fquatinv(q);
end
time_end = clock;
quatinv_average = etime(time_end,time_begin)*1000/times


time_begin = clock;
for i=1:times
    q_normalized = quatnormalize(q);
end
time_end = clock;
quatnormalize_average = etime(time_end,time_begin)*1000/times

time_begin = clock;
for i=1:times
    q_normalized = fquatnormalize(q);
end
time_end = clock;
quatnormalize_average = etime(time_end,time_begin)*1000/times


time_begin = clock;
for i=1:times
    q_prod = quatmultiply(q1, q2);
end
time_end = clock;
quatmultiply_average = etime(time_end,time_begin)*1000/times

time_begin = clock;
for i=1:times
    q_prod = fquatmultiply(q1, q2);
end
time_end = clock;
quatmultiply_average = etime(time_end,time_begin)*1000/times


time_begin = clock;
for i=1:times
    p_new = quatrotate(q, p');
end
time_end = clock;
quatrotate_average = etime(time_end,time_begin)*1000/times

time_begin = clock;
for i=1:times
    p_new = fquatrotate(q, p);
end
time_end = clock;
quatrotate_average = etime(time_end,time_begin)*1000/times






